Replace jemalloc with mimalloc config.toml option

jq-rs opened the pull request:
* This is a test PR for swapping jemalloc to mimalloc.
* The performance results in my own environment are promising.
* This is not intended to be merged upstream as such, just to provide initial performance results which can lead to further discuss what to do next.

Edit: After the results, changed mimalloc to be introduced as its own option in `config.toml.example` and added support via 22f306c. Changed title accordingly and dropped WIP status.

Edit: After review comments, changed mimalloc to replace jemalloc as config.toml option.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @Mark-Simulacrum (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://rustc-dev-guide.rust-lang.org/contributing.html) for more information.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1838177938) [(plain)](https://github.com/rust-lang/rust/commit/6fc64b2ba6650a0aef60ef3048977ed81eb30c2c/checks/1838177938/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
skip untracked path cpu-usage.csv during rustfmt invocations
skip untracked path src/doc/book/ during rustfmt invocations
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2018" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc/src/main.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
Diff in /checkout/compiler/rustc/src/main.rs at line 17:
         static _F2: unsafe extern "C" fn(*mut *mut c_void, usize, usize) -> c_int =
             libmimalloc_sys::mi_posix_memalign;
         #[used]
-        static _F3: unsafe extern "C" fn(usize, usize) -> *mut c_void = libmimalloc_sys::mi_aligned_alloc;
+        static _F3: unsafe extern "C" fn(usize, usize) -> *mut c_void =
+            libmimalloc_sys::mi_aligned_alloc;
         #[used]
         static _F4: unsafe extern "C" fn(usize) -> *mut c_void = libmimalloc_sys::mi_malloc;
         #[used]
Diff in /checkout/compiler/rustc/src/main.rs at line 24:
-        static _F5: unsafe extern "C" fn(*mut c_void, usize) -> *mut c_void = libmimalloc_sys::mi_realloc;
+        static _F5: unsafe extern "C" fn(*mut c_void, usize) -> *mut c_void =
+            libmimalloc_sys::mi_realloc;
         #[used]
         static _F6: unsafe extern "C" fn(*mut c_void) = libmimalloc_sys::mi_free;
failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test --stage 2 src/tools/tidy
Build completed unsuccessfully in 0:00:17
```

</details>

jq-rs commented:
We probably would need bors for benchmarks?

Aaron1011 commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 35d93523681994e8fad7a3c4886e7072e6344040 with merge 86f3a470dd7fa102be01dc6d84ecc6920f8af32a...
<!-- homu: {"type":"TryBuildStarted","head_sha":"35d93523681994e8fad7a3c4886e7072e6344040","merge_sha":"86f3a470dd7fa102be01dc6d84ecc6920f8af32a"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1840075583)
Build commit: 86f3a470dd7fa102be01dc6d84ecc6920f8af32a (`86f3a470dd7fa102be01dc6d84ecc6920f8af32a`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1840075583"},"merge_sha":"86f3a470dd7fa102be01dc6d84ecc6920f8af32a"} -->

rust-timer commented:
Queued 86f3a470dd7fa102be01dc6d84ecc6920f8af32a with parent f9435f4c92651d67d5dbaba13c5606c4c4fc1327, future [comparison URL](https://perf.rust-lang.org/compare.html?start=f9435f4c92651d67d5dbaba13c5606c4c4fc1327&end=86f3a470dd7fa102be01dc6d84ecc6920f8af32a).

rust-timer commented:
Finished benchmarking try commit (86f3a470dd7fa102be01dc6d84ecc6920f8af32a): [comparison url](https://perf.rust-lang.org/compare.html?start=f9435f4c92651d67d5dbaba13c5606c4c4fc1327&end=86f3a470dd7fa102be01dc6d84ecc6920f8af32a).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf

nagisa commented:
While there are major decrease in the instruction counts (yay!), the wall-clock time does not share a similarly stark change (though bootstrap times are all universally better) and there's a large increase in max-rss on some benchmarks. Pretty interesting tradeoff, overall.

jq-rs commented:
Pretty good results, I think! How well the benchmarks measure multi-threaded performance? I believe I see even better results in such a setup.

If we want to add mimalloc support, maybe the best option would be to add mimalloc as a new config.toml option? Mimalloc is in active development, so results may improve even more in the future.

jyn514 commented:
> the wall-clock time does not share a similarly stark change

Hmm, why do you say that? I see up to -11.9% decreases, and large decreases across the board.

![image](https://user-images.githubusercontent.com/23638587/107131631-3ea82980-68a6-11eb-963e-7b0ca631b1e1.png)

jyn514 commented:
> Pretty good results, I think! How well the benchmarks measure multi-threaded performance? I believe I see even better results in such a setup.

Not at all currently I think, `cfg(parallel_compiler)` is off by default (and isn't currently seeing active work: https://github.com/rust-lang/compiler-team/blob/master/content/minutes/design-meeting/2020-05-29-Roadmap-2020-2021.md#key-notes-and-points-from-the-meeting-itself).

jq-rs commented:
Added mimalloc option support for config.toml for a reference via 22f306c. Removed misleading OSX mention from config.toml.example comments from jemalloc.

Mark-Simulacrum commented:
Is there a reason this intends to *add* mimalloc, instead of replacing jemalloc? I'm not seeing clear reasons we would want to have both as options. I am also not seeing any huge increases in max-rss - 20-25% is pretty typical variability I think.

jq-rs commented:
I was thinking the following path to mimalloc as the new default:

1. Add mimalloc as a new option in the config.toml
2. Change compiler default to mimalloc in tooling
3. Deprecate jemalloc config.toml option (needs some release note docs)
4. Remove jemalloc config.toml option

It may be confusing for rustc users to lose their explicit config.toml option, so that is why a deprecation step is probably useful. Changing all CI and tools to use mimalloc is a separate thing, and thus I think it is better to do in a separate PR.

So yes, I agree that mimalloc should be the only default in the end.

I can do these steps (but I may need some help to catch all things needed in every step).

jyn514 commented:
@jq-rs I'm happy to mentor the steps :) adding the option (1) should be pretty simple, (3) just needs a note in `src/boostrap/CHANGELOG.md`. I'm a little confused by (2) though - AFAIK jemalloc isn't the default on any platform, are you suggesting mimalloc should replace the system allocator? That does sound nice but I'd want to see why jemalloc was turned off by default first.

jq-rs commented:
@jyn514 Thanks!
1) should be this PR, right?
2) I think jemalloc is the default in CI (and bors) by rustc? So we just should swap mimalloc there. The idea is not to replace system allocator, just allow to use mimalloc with config.toml-setting.

jyn514 commented on src/bootstrap/config.rs:
```diff
@@ -849,6 +851,7 @@ impl Config {
             set(&mut config.codegen_tests, rust.codegen_tests);
             set(&mut config.rust_rpath, rust.rpath);
             set(&mut config.jemalloc, rust.jemalloc);
+            set(&mut config.mimalloc, rust.mimalloc);
```
What happens if both are set? Will cargo give a compile error? If not, can you add a check in bootstrap?

jq-rs replied:
It will give the following compilation error, if both are enabled: <pre><code>error: Package `rustc-main v0.0.0 (/home/jkettune/rust/compiler/rustc)` does not have the feature `jemallocmimalloc`</code></pre>

It is not a that nice error, but maybe it will do for the transition period.

lqd replied:
I believe this is due to https://github.com/rust-lang/rust/pull/81782/files?file-filters%5B%5D=.example&file-filters%5B%5D=.rs&file-filters%5B%5D=.toml#diff-b46049dee2a9f04cac5ccc2e6820ecf7ef17d996c6f7258c85759cd48937b1f9R567-R572 which will concatenate both feature flags

jyn514 reviewed (approved):
r=me with nit fixed. This is awesome :heart:

jyn514 commented:
@bors r+

bors commented:
:pushpin: Commit 22f306c6508af55ac66e25e432449431808b9890 has been approved by `jyn514`

<!-- @bors r=jyn514 22f306c6508af55ac66e25e432449431808b9890 -->
<!-- homu: {"type":"Approved","sha":"22f306c6508af55ac66e25e432449431808b9890","approver":"jyn514"} -->

bors commented:
:hourglass: Testing commit 22f306c6508af55ac66e25e432449431808b9890 with merge deb945bca06f39d0eddaaa92bb0e49e5688a39c3...
<!-- homu: {"type":"BuildStarted","head_sha":"22f306c6508af55ac66e25e432449431808b9890","merge_sha":"deb945bca06f39d0eddaaa92bb0e49e5688a39c3"} -->

Mark-Simulacrum commented:
I guess I don't agree with this series of steps, I don't think we should keep an option we're not testing and that we're intending to deprecate. I would also like to see a crater run on this PR before we merge it.

@bors r-

jq-rs commented:
@Mark-Simulacrum It is possible to have all four steps within this PR. With a quick peek, in addition to these changes we probably need to:
1. Drop `jemalloc` support from this PR
2. Update `src/ci/github-actions/ci.yml` for mimalloc and drop jemalloc
3. Update Docker-files (e.g. `src/ci/docker/host-x86_64/dist-x86_64-linux/Dockerfile`) for mimalloc and drop jemalloc
4. Document the change to `src/bootstrap/CHANGELOG.md`

Is there something else?

bors commented:
:boom: Test timed out
<!-- homu: {"type":"TimedOut"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/1887965431) [(plain)](https://github.com/rust-lang-ci/rust/commit/deb945bca06f39d0eddaaa92bb0e49e5688a39c3/checks/1887965431/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

Mark-Simulacrum commented:
That sounds good to me. Once we're ready I'll also kickoff a try build and crater run; I wouldn't expect any problems but I think this is a significant enough change that it makes sense to check.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1901283428) [(plain)](https://github.com/rust-lang/rust/commit/d4e65c60386cce4eac8614ec9ad2dbb89a574365/checks/1901283428/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
   Compiling thiserror v1.0.20
   Compiling yaml-merge-keys v0.4.1
   Compiling expand-yaml-anchors v0.1.0 (/checkout/src/tools/expand-yaml-anchors)
    Finished release [optimized] target(s) in 6.47s
error: .github/workflows/ci.yml is not up to date; please run `x.py run src/tools/expand-yaml-anchors`.


caused by: src/ci/github-actions/ci.yml and .github/workflows/ci.yml are different
command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/expand-yaml-anchors" "check" "/checkout"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap --stage 2 test src/tools/expand-yaml-anchors
Build completed unsuccessfully in 0:00:06
```

</details>

jq-rs commented:
@Mark-Simulacrum All right, `mimalloc` changes are ready for try and crater rounds, if they look good otherwise.

Mark-Simulacrum commented:
@bors try @rust-timer queue

I'll kickoff crater once try is done.

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 388bd5e31394477ad0ecb067f36bc7414963dcf2 with merge 15a786332e2f0dd343b226ab6bb8ec028a350f73...
<!-- homu: {"type":"TryBuildStarted","head_sha":"388bd5e31394477ad0ecb067f36bc7414963dcf2","merge_sha":"15a786332e2f0dd343b226ab6bb8ec028a350f73"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1904141458)
Build commit: 15a786332e2f0dd343b226ab6bb8ec028a350f73 (`15a786332e2f0dd343b226ab6bb8ec028a350f73`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1904141458"},"merge_sha":"15a786332e2f0dd343b226ab6bb8ec028a350f73"} -->

rust-timer commented:
Queued 15a786332e2f0dd343b226ab6bb8ec028a350f73 with parent 9503ea19edbf01b9435e80e17d60ce1b88390116, future [comparison URL](https://perf.rust-lang.org/compare.html?start=9503ea19edbf01b9435e80e17d60ce1b88390116&end=15a786332e2f0dd343b226ab6bb8ec028a350f73).

Mark-Simulacrum commented:
@craterbot check

rust-timer commented:
Finished benchmarking try commit (15a786332e2f0dd343b226ab6bb8ec028a350f73): [comparison url](https://perf.rust-lang.org/compare.html?start=9503ea19edbf01b9435e80e17d60ce1b88390116&end=15a786332e2f0dd343b226ab6bb8ec028a350f73).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf

davidtgoldblatt commented:
BTW, coming from jemalloc upstream (where this issue got linked) -- this poked me to take a look at rustc performance with jemalloc, and looking at the profiles, it seems like Rust isn't using the sized deallocation APIs (e.g. it throws away `layout` in the dealloc implementation and just calls free, rather than doing what e.g. jemallocator does and passing it to the allocator).

This likely explains the perf difference -- we more or less regard `free` as a slow path these days, and optimize heavily assuming that the large majority of deallocations pass us the size information (as they should in most Rust/C++ programs when properly configured).

(No idea if rustc has special considerations that make this infeasible or something, and we wouldn't take it personally if it's easier to switch allocators rather than make the Rust allocation code config-dependent; just explaining why upstream jemalloc is probably going to be unmotivated to make this pattern fast).

Mark-Simulacrum commented:
Interesting! I'm not aware of reasons for us to not do so, it definitely sounds like we should compare against that configuration. Thanks for taking a look.

davidtgoldblatt commented:
(There's the other issue that current "official jemalloc release" is shamefully old, but that's not something I think rustc can work around).

LifeIsStrange commented:
In addition to the latest official jemalloc release dating back to 2019, here jemalloc-sys 0.3 is being used which use a jemalloc version from 2018!

Experimenting with either https://crates.io/crates/tikv-jemalloc-sys
https://crates.io/crates/tikv-jemallocator
Would be better as they effectively use the latest stable

Mark-Simulacrum commented:
@jq-rs would you be up for doing the legwork to update our jemalloc and otherwise bring it more in line with the best possible we can do from jemalloc, before we evaluate switching to mimalloc? I think it makes sense to stick to the same general implementation rather than bouncing around too much if we can upgrade and use the sized deallocation functions.

jq-rs commented:
@Mark-Simulacrum How about if we change the `config.toml` option to `extmalloc`? Then we do not need to bounce with the option itself and can change the underlying library easily based on measurements later on.

Mark-Simulacrum commented:
I don't think it's worth changing the config option to something generic - I'd rather change it to a specific allocator if we decide to switch.

jq-rs commented:
The recent [benchmarks](https://github.com/microsoft/mimalloc#Performance) of `mimalloc` are quite convincing, so I would not expect `jemalloc` to be _faster_. We can measure it to be sure, of course.

I can push a separate test PR with [tikv-jemalloc-sys](https://crates.io/crates/tikv-jemalloc-sys) with latest stable `jemalloc` for reference. Can someone pinpoint where exactly we would need the free-changes?

jq-rs commented:
I checked briefly is there a way to pass `layout.size()` for the `free` call. As far as I can understand, we would need to call non-standard `sdallocx` API as the free API, similar to jemallocator dealloc(). It may be difficult, as it is not part of `libc` standard API. The global allocator functionality is built on the standard libc API.

Now unix alloc.rs just calls
``` 
    #[inline]
    unsafe fn dealloc(&self, ptr: *mut u8, _layout: Layout) {
        libc::free(ptr as *mut libc::c_void)
    }
```

Mark-Simulacrum commented:
It should be fine for ad-hoc profiling to modify that function, though we'd definitely need to figure out an alternative route before landing. I imagine mimalloc would also benefit from getting layout info on deallocation?

jq-rs commented:
By difficult I mean perhaps infeasible. The compiler allocator system is currently designed for _drop-in_ allocator use, as we need to support system allocators on some architectures. It seems reasonable and elegant design to me where function override can be leveraged. Any _drop-in_ allocator has to support these paths _as well as they can_ to be able to compete with other drop-in allocators. This is a fair game as such.

For ad-hoc testing we can change the free call to non-standard version for rustc, but I don't think it provides a path forward here. AFAIK it does not help with LLVM override either. So, if we do it, we should do it for completeness, and perhaps for curiosity.

pcwalton commented:
It's disappointing if we've boxed ourselves into a corner where we can't use sized deallocation APIs anymore. Many years ago we actually made a design decision to used sized deallocation wherever possible. I think we should try hard to be able to use sized deallocation again, because I don't think we ever officially went back on that decision.

pcwalton commented:
cc @alexcrichton -- did we actually decide to drop sized deallocation at any point? :(

alexcrichton commented:
The problem, IIRC, has nothing to do with Rust-the-language or whether we've boxed ourselves out of this, but rather it has everything to do with how we distribute and build the compiler itself. The compiler is built as `rustc` which links to `librustc_driver.so` which links to `libstd.so`. The standard library must have an allocator picked by the time we link it and we can't really reasonably say that all users of `libstd.so` must use whatever works for rustc. As a result we pick the system `malloc`/`free` symbols for `libstd.so`, and then overriding the allocator in rustc involves somehow overriding the system allocator (which is pretty platform-specific AFAIK).

This means that rustc, due to how its built, has no way of using `#[global_allocator]` in Rust which would allow routing all Rust allocation/deallocation to custom APIs (e.g. a sized deallocation API). This has historically been coupled with the fact that there was no way (or no one knew a way) to hook up LLVM into sized deallocation APIs.

To leverage sized deallocation I think what needs to happen is:

* First probably measure the impact to see whether it's at all worth it.
* Decide that https://github.com/rust-lang/rust/pull/56986 should "be reverted" in that we want to force all tools using `librustc_driver.so` (e.g. RLS, rustdoc, rustc, etc) to use the allocator we pick. https://github.com/rust-lang/rust/pull/56986 originally concluded that jemalloc was not appropriate for the RLS at the time.
* Decouple `librustc_driver.so` from `libstd.so`. I forget if it's just legacy reasons why we still do this or whether there's a technical reason. It's probably more of the former. This is likely a large-ish undertaking.
* Use `#[global_allocator]` in `librustc_driver.so`, hooking into custom allocator API.
* Either make sure LLVM's allocations are routed to the custom allocator used via some platform-specific means or figure out how to route LLVM's allocations to the same custom allocator API.

AFAIK none of this is a light undertaking. Depending on the allocator is also may only benefit one platform instead of all.

jq-rs commented:
@Mark-Simulacrum Regarding your earlier question, `mimalloc` does not benefit from sized deallocation, `free` is the only fast path as can be seen [here](https://github.com/microsoft/mimalloc/blob/master/src/alloc.c#L571-L575).

bors commented:
:umbrella: The latest upstream changes (presumably #82896) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

jq-rs commented:
`jemalloc` just got macOS fixes in, which is grand. `mimalloc` would need similar fixes for macOS, so they would need to be included for this PR before merging. But where are we with this one, are these macOS fixes worth working for? 

In addition to the above, my view is that the discussion so far has all the input right and we would just need the decision shall we switch allocator at this point. `mimalloc` seems perhaps faster against the latest `jemalloc` dev `free` path and is faster with rustc compared to latest stable 5.2.1 (see #82389 measurements). Sized allocation seems a bit faster approach on some allocators (especially `jemalloc` and `tcmalloc`) but that does not apply for `mimalloc` at the moment.

How I see this is that if we decide to stay with the drop-in allocator approach for now or for some time, `mimalloc` is certainly worth considering. If we plan to use sized deallocation in the near future, then changing to `mimalloc` now would perhaps not be worth it.

jq-rs commented:
And for details _why_ I think "`mimalloc` seems perhaps faster against the latest `jemalloc` dev `free` path" is this `mimalloc-bench` between `mi` 1.7.0 and `je` dev (73ca4b8) where in most tests mimalloc is faster. It would be better to test this with the real application, but as it is difficult, this is the best I could do. The larson-sized test is an extension I did for the benchmarks to have at least one measurement data point for sized deallocation (the benchmark can be found [here](https://github.com/daanx/mimalloc-bench/pull/25)):
```
# benchmark allocator elapsed rss user sys page-faults page-reclaims

cfrac mi 08.19 3104 8.18 0.00 0 362
cfrac je 08.31 4836 8.29 0.00 0 470

espresso mi 05.97 4096 5.93 0.02 0 678
espresso je 06.17 5212 6.12 0.03 0 648

barnes mi 04.34 58276 4.31 0.02 0 15604
barnes je 04.34 60276 4.30 0.02 0 15732

leanN mi 34.29 502904 121.55 1.65 0 371062
leanN je 34.44 464624 120.82 1.56 0 349706

redis mi 7.467 30064 2.92 0.82 0 9830
redis je 9.077 33672 3.91 0.64 0 8406

alloc-test1 mi 04.08 12744 4.06 0.01 1 7122
alloc-test1 je 04.29 12296 4.27 0.00 0 6802

alloc-testN mi 04.76 14552 70.95 0.02 0 5792
alloc-testN je 05.76 15440 77.03 0.03 0 9390

larsonN mi 2.889 312180 199.19 0.56 0 80847
larsonN je 2.962 256788 197.21 1.17 0 140290

larsonN-sized mi 2.829 318884 199.01 0.60 0 82602
larsonN-sized je 2.758 259904 196.79 1.34 0 148061

sh6benchN mi 00.16 288940 3.27 0.22 0 71998
sh6benchN je 00.20 284776 4.50 0.33 0 73431

sh8benchN mi 00.38 208100 7.72 0.15 0 51784
sh8benchN je 00.65 171168 19.64 0.26 0 51550

xmalloc-testN mi 0.457 134344 190.65 1.69 1 60977
xmalloc-testN je 1.804 110368 165.07 17.96 0 57251

cache-scratch1 mi 02.53 3972 2.52 0.00 0 265
cache-scratch1 je 02.52 4588 2.51 0.00 0 345

cache-scratchN mi 00.17 4608 4.80 0.00 0 467
cache-scratchN je 00.16 6724 5.39 0.00 0 992

mstressN mi 06.67 2583856 64.01 3.57 0 648395
mstressN je 13.00 695004 33.78 39.55 0 8733729

rptestN mi 3.529 296848 27.70 5.77 1 73968
rptestN je 4.801 201188 37.23 5.94 0 105776
```

joshtriplett commented:
@alexcrichton wrote:
> Either make sure LLVM's allocations are routed to the custom allocator used via some platform-specific means or figure out how to route LLVM's allocations to the same custom allocator API.

As long as we never assume we can call our free function on something that LLVM allocated, couldn't we just make LLVM support calling an allocator other than the system allocator, and then assume that if someone uses a shared LLVM, they want to use whatever allocator that LLVM is built to use? That way we don't have to do any kind of interposition on the system allocator at all.

jq-rs commented:
As this one needs macOS updates, closing this for now. After #83152 is merged, this can be compared again with macOS support. Thanks all!

nothingnesses commented:
@jq-rs  #83152 has been merged. Any updates on this?

jq-rs commented:
@nothingnesses AFAIK macOS is not supported by `mimalloc` fully, so this is not progressing.

jq-rs commented:
And I verified this, the macOS override part is not there yet for `mimalloc` in v1.7.2 at least. If some macOS specialist wants to take a look, I am happy to give a patch for the other replace parts.

fee1-dead commented:
It looks like they improved macOS overriding support in v1.7.3: https://github.com/microsoft/mimalloc/commit/c47de7eb9090cf8c0a5f307a74437cf927a361be
